### PR TITLE
Add minicpm5 tool call parser

### PR DESCRIPTION
**Pull request by zhangtao2-1**
## Overview

Add MiniCPM5 tool call support for `llama-server`.

MiniCPM5 outputs tool calls as XML (`<function name="..."><param name="...">...</param></function>`), not JSON. This PR adds a `peg-minicpm5` parser to detect the template, parse tool calls (including streaming), normalize common output quirks, and expose OpenAI-compatible `tool_calls`.

Also fixes a peg mapper streaming bug (use tool index instead of dangling pointer) and adds Jinja `min`/`max` filters needed by the MiniCPM5 template.

## Test plan

- [x] `test-chat-peg-parser-minicpm5`
- [x] `test-jinja` (`min` / `max`)
- [x] Manual test with MiniCPM5 GGUF + tools on `llama-server`

### CODE REVIEW

**Comment by zhangtao2-1** [comment]
@CISC would you mind taking a look at this PR? MiniCPM5 tool call parsing — would appreciate your review.

File: common/jinja/value.cpp [hunk 1]
```diff
@@ -1089,6 +1089,34 @@ const func_builtins & value_array_t::get_builtins() const {
...
+            }
+            value result = arr[0];
+            for (size_t i = 1; i < arr.size(); ++i) {
+                if (value_compare(arr[i], result, value_compare_op::lt)) {
+                    result = arr[i];
+                }
+            }
+            return result;
+        }},
+        {"max", [](const func_args & args) -> value {
```
Comment (CISC) [inline common/jinja/value.cpp:1133]: Add `case_sensitive` arg and a `FIXME` as in `sort`.

Add `attribute` arg and either implement it, or throw `not_implemented_exception`.

File: tests/test-jinja.cpp [hunk 2]
```diff
@@ -1534,6 +1534,18 @@ static void test_array_methods(testing & t) {
         "6"
     );
 
+    test_template(t, "array|min",
+        "{{ [tool_calls_count, tool_sep_count]|min }}",
+        {{"tool_calls_count", 2}, {"tool_sep_count", 1}},
+        "1"
+    );
+
+    test_template(t, "array|max",
```
Comment (CISC) [inline tests/test-jinja.cpp:1567]: Add `attribute` tests.

**Comment by CISC** [comment]
> @CISC would you mind taking a look at this PR? MiniCPM5 tool call parsing — would appreciate your review.

I'll review the `jinja` part, leaving the parser to @ggml-org/llama-common

File: common/jinja/value.cpp (hunk 1 above)
Comment (zhangtao2-1) [reply common/jinja/value.cpp:1133]: Thanks for the review! Will update min/max to match sort's case_sensitive/attribute API and add attribute tests.

File: common/chat-peg-parser.cpp [hunk 3]
```diff
@@ -211,7 +211,18 @@ common_peg_parser common_chat_peg_builder::tag_with_safe_content(const std::stri
...
+}
+
+common_chat_tool_call * common_chat_peg_mapper::active_tool() {
+    if (committed_tool_idx.has_value()) {
+        return &result.tool_calls.at(committed_tool_idx.value());
+    }
+    if (pending_tool_call.has_value()) {
+        return &pending_tool_call.value();
+    }
+    return nullptr;
```
Comment (aldehir) [inline common/chat-peg-parser.cpp]: If it's not broke, don't fix it.

File: common/chat-peg-parser.cpp [hunk 4]
```diff
@@ -1003,3 +1021,69 @@ void common_chat_peg_gemma4_mapper::visit(const common_peg_ast_arena & arena, co
         visit(arena, child_id);
     }
 }
+
+common_peg_parser common_chat_peg_builder::minicpm5_xml_tool_calls(const ordered_json & tools,
```
Comment (aldehir) [inline common/chat-peg-parser.cpp]: Inline in the chat param init for the model template.

File: common/chat-peg-parser.cpp [hunk 5]
```diff
@@ -1003,3 +1021,69 @@ void common_chat_peg_gemma4_mapper::visit(const common_peg_ast_arena & arena, co
...
+        const std::string name = function.at("name");
+        ordered_json params = function.contains("parameters") ? function.at("parameters") : ordered_json::object();
+
+        auto args = eps();
+        if (params.contains("properties") && params.at("properties").is_object() &&
+            !params.at("properties").empty()) {
+            auto arg_choice = choice();
+            for (const auto & el : params.at("properties").items()) {
+                const std::string & prop_name = el.key();
+                const std::string & prop_type = el.value().value("type", "string");
```
Comment (aldehir) [inline common/chat-peg-parser.cpp]: Please use the established patterns in the repo for determining if a type is a string. See `common_schema_info`.

File: common/chat.cpp [hunk 6]
```diff
@@ -2253,6 +2256,153 @@ static void func_args_not_string(json & messages) {
...
+    static const std::string IM_END = "<|im_end|>";
+    if (out.size() >= IM_END.size() &&
+        out.compare(out.size() - IM_END.size(), IM_END.size(), IM_END) == 0) {
+        out.erase(out.size() - IM_END.size());
+        while (!out.empty() && (out.back() == '\n' || out.back() == ' ')) {
+            out.pop_back();
+        }
+    }
+
+    return out;
```
Comment (aldehir) [inline common/chat.cpp]: All of this should be defined in the parser. None of this looks like it needs to be done post process.

File: common/chat.cpp [hunk 7]
```diff
@@ -2253,6 +2256,153 @@ static void func_args_not_string(json & messages) {
...
+    static const std::string SP_NL    = "\xC1\x8A"; // U+010A
+
+    for (size_t pos = 0; (pos = out.find(SP_SPACE, pos)) != std::string::npos;) {
+        out.replace(pos, SP_SPACE.size(), " ");
+        pos += 1;
+    }
+    for (size_t pos = 0; (pos = out.find(SP_NL, pos)) != std::string::npos;) {
+        out.replace(pos, SP_NL.size(), "\n");
+        pos += 1;
+    }
```
Comment (aldehir) [inline common/chat.cpp]: This seems easy enough to do in a single pass.

File: common/chat.cpp [hunk 8]
```diff
@@ -2253,6 +2256,153 @@ static void func_args_not_string(json & messages) {
...
+    data.supports_thinking = true;
+    data.preserved_tokens  = {
+        "<function",
+        "<param",
+        "</function>",
+        "</param>",
+        "<think>",
+        "</think>",
+        "<|im_start|>",
+        "<|im_end|>",
```
Comment (aldehir) [inline common/chat.cpp]: Remove the EOS token and you won't need to trim it from the output.

File: common/chat.cpp [hunk 9]
```diff
@@ -2253,6 +2256,153 @@ static void func_args_not_string(json & messages) {
...
+
+        auto assistant_prefix = p.literal(GEN_PROMPT);
+        auto thinking_prefix  = inputs.enable_thinking ?
+            p.literal("<think>\n") :
+            p.literal("<think>\n\n</think>\n\n");
+
+        auto reasoning = p.eps();
+        if (extract_reasoning && inputs.enable_thinking) {
+            reasoning = p.reasoning(p.until_one_of(TOOL_START_MARKERS)) + p.optional(p.literal("\n")) +
+                        p.optional(p.literal(THINK_END) + p.optional(p.literal("\n\n")));
```
Comment (aldehir) [inline common/chat.cpp]: Too many optionals. This can be a choice with two branches: one that completes a thought with the end thinking tag and one that preempts thinking with the start of a tool call. Use `p.space()` to consume whitespace.

File: common/chat.cpp [hunk 10]
```diff
@@ -2253,6 +2256,153 @@ static void func_args_not_string(json & messages) {
...
+        }
+
+        auto suffix = p.optional(p.literal("<|im_end|>") + p.optional(p.literal("\n")));
```
Comment (aldehir) [inline common/chat.cpp]: Should not be needed.

File: common/chat.cpp [hunk 11]
```diff
@@ -2253,6 +2256,153 @@ static void func_args_not_string(json & messages) {
...
+
+    data.parser = parser.save();
+
+    // MiniCPM5 tool calls are parsed post-hoc via peg-minicpm5 (XML output).
+    // Do not attach JSON-schema GBNF here — it is invalid for this format and
+    // can destabilize llama-server. SGLang/vLLM use parsers only for MiniCPM5.
+    data.grammar.clear();
+    data.grammar_lazy     = false;
+    data.grammar_triggers = {};
+    (void) include_grammar;
```
Comment (aldehir) [inline common/chat.cpp]: Even if you don't use the JSON schema to grammar implementation, the PEG parser will compose GBNF rules to enforce tool call structure. Please don't deviate from established patterns.

File: common/chat.cpp [hunk 12]
```diff
@@ -2563,9 +2719,13 @@ common_chat_msg common_chat_peg_parse(const common_peg_arena &          src_pars
...
 
+    const std::string normalized_input = params.format == COMMON_CHAT_FORMAT_PEG_MINICPM5 ?
+        common_chat_normalize_minicpm5_output(input) :
+        input;
+
     const std::string effective_input = params.generation_prompt.empty()
-        ? input
-        : params.generation_prompt + input;
+        ? normalized_input
+        : params.generation_prompt + normalized_input;
```
Comment (aldehir) [inline common/chat.cpp]: Normalization should occur in a dedicated mapper. Not here.

File: tests/CMakeLists.txt [hunk 13]
```diff
@@ -195,6 +195,7 @@ if (NOT WIN32 OR NOT BUILD_SHARED_LIBS)
 endif()
 
 llama_build_and_test(test-chat-peg-parser.cpp peg-parser/simple-tokenize.cpp)
+llama_build_and_test(test-chat-peg-parser-minicpm5.cpp)
```
Comment (aldehir) [inline tests/CMakeLists.txt]: To reiterate, do not deviate from established patterns. Chat template tests are located under `test-chat.cpp`.

**Review by aldehir — CHANGES_REQUESTED** [review]
Thanks for the PR. The code present deviates too much from what is already established. I don't really see anything that would necessitate the need for these deviations.

I'll let @pwilkin weigh in if he thinks a dedicated parser is necessary or if this can be handled by the autoparser with some tweaks.

**Comment by zhangtao2-1** [comment]
@aldehir Thanks. Understood on aligning with established patterns — I'll refactor accordingly. Also happy to wait for @pwilkin's input on dedicated parser vs. autoparser before going too far in either direction.

**Comment by pwilkin** [comment]
Can you add the chat template here? Then we'll be able to just test it with the autoparser and see how much works out of the box and what needs to be possibly fixed.

**Comment by zhangtao2-1** [comment]
@pwilkin @CISC @aldehir 
Pushed 5187d06 with review fixes: align with established patterns (inline PEG, common_schema_info, dedicated minicpm5_mapper, GBNF restored, no generic mapper changes), jinja min/max API + tests, MiniCPM5 template fixture for autoparser experiments, tests in test-chat.cpp. test-chat / test-jinja / test-chat-peg-parser pass. PTAL — especially on dedicated parser vs autoparser.

File: common/jinja/value.cpp [hunk 14]
```diff
@@ -1089,6 +1089,50 @@ const func_builtins & value_array_t::get_builtins() const {
             std::reverse(arr.begin(), arr.end());
             return is_val<value_tuple>(val) ? mk_val<value_tuple>(std::move(arr)) : mk_val<value_array>(std::move(arr));
         }},
+        {"min", [](const func_args & args) -> value {
+            args.ensure_count(1, 4);
+            args.ensure_vals<value_array>();
+            value attribute   = args.get_kwarg_or_pos("attribute",      1);
+            value val_case    = args.get_kwarg_or_pos("case_sensitive", 2);
```
Comment (CISC) [inline common/jinja/value.cpp]: Swap places, `case_sensitive` is first positional arg.

File: common/jinja/value.cpp [hunk 15]
```diff
@@ -1089,6 +1089,50 @@ const func_builtins & value_array_t::get_builtins() const {
...
+            if (!attribute->is_undefined()) {
+                throw not_implemented_exception("min: attribute not implemented");
+            }
+            // FIXME: min is currently always case sensitive
+            (void) val_case;
+            const auto & arr = args.get_pos(0)->as_array();
+            if (arr.empty()) {
+                throw raised_exception("min() arg is an empty sequence");
+            }
```
Comment (CISC) [inline common/jinja/value.cpp:1124]: An empty sequence is allowed, it simply returns `undefined`.

File: tests/test-jinja.cpp [hunk 16]
```diff
@@ -1534,6 +1534,49 @@ static void test_array_methods(testing & t) {
...
+        "{{ [tool_calls_count, tool_sep_count]|max }}",
+        {{"tool_calls_count", 2}, {"tool_sep_count", 1}},
+        "2"
+    );
+
+    test_template(t, "array|min attribute not implemented",
```
Comment (CISC) [inline tests/test-jinja.cpp]:
```suggestion
    test_template(t, "array|min attribute",
```

File: tests/test-jinja.cpp [hunk 17]
```diff
@@ -1534,6 +1534,49 @@ static void test_array_methods(testing & t) {
...
+        "{{ items|min(attribute='x') }}",
+        {{"items", json::array({
+            json({{"x", 2}}),
+            json({{"x", 1}}),
+        })}},
+        ""
+    );
+
+    test_template(t, "array|max attribute not implemented",
```
Comment (CISC) [inline tests/test-jinja.cpp]:
```suggestion
    test_template(t, "array|max attribute",
```

File: tests/test-jinja.cpp [hunk 18]
```diff
@@ -1534,6 +1534,49 @@ static void test_array_methods(testing & t) {
...
+        "{{ items|min(case_sensitive=false) }}",
+        {{"items", json::array({"B", "a", "A"})}},
+        "A"
+    );
+
+    test_template(t, "array|max case_sensitive",
+        "{{ items|max(case_sensitive=false) }}",
+        {{"items", json::array({"B", "a", "A"})}},
+        "a"
+    );
```
Comment (CISC) [inline tests/test-jinja.cpp]:
```suggestion
```
Just remove these for now, CI will fail, besides the test should test both `true` and `false` when eventually supported as `false` should have been the default.

Comment (zhangtao2-1) [reply tests/test-jinja.cpp]: DONE

**Comment by zhangtao2-1** [comment]
Hi @pwilkin @aldehir 
When you have a moment, could you please review my PR? Thanks!

**Comment by lexasub** [comment]
@pwilkin review pls

**Comment by lexasub** [comment]
image
image
@zhangtao2-1, hello, i fetched you branch (and rebuild) and try test tool calling (mistral vibe, opencode) on your template
./bin/llama-server --chat-template-file ../models/templates/MiniCPM5-1B.jinja  --host 0.0.0.0 --jinja -fa on --port 1113  --ctx-size 200000 --model ./MiniCPM5-1B-Q4_K_M.gguf  (qwopus - it name in mistral vibe config)

**Comment by zhangtao2-1** [comment]
Hi @lexasub 
Thanks for the repro — the two screenshots point to two different issues:

1) HTTP 500 (Failed to parse tool call arguments as JSON)
On multi-turn requests, malformed/truncated tool_calls.arguments in history (e.g. {"url") caused func_args_not_string() to throw → 500. Fixed on this branch: parse failures now log a warning and fall back to {} instead of aborting. Added a multi-turn regression test in test-chat.cpp.

2) <command_utilization> + H5H5... repetition
Not a MiniCPM5 XML parser issue — the model output mistral vibe/opencode’s format, not <function name="..."><param ...>. The H5 loop is small-model repetition (1B Q4 under agent load).

For agent testing: use --reasoning-budget 256, "enable_thinking": false for tool tasks, max_tokens ≥ 1024, and prefer F16 if possible.

The 500 on follow-up turns should be fixed on the latest branch — would appreciate a retest. Thanks!

**Comment by pwilkin** [comment]
Sorry for the delay, I had a busy time recently and was AFK all weekend, could only access comments.

From what I determined, the specifics of the MiniCPM5 template format are that it doesn't have a tool call marker per se - its tool calls start directly with the function marker, but it has a call separator marker. I think the cleanest solution will be to add support for this type of behavior in the autoparser, if that fails, we can fall back to the dedicated parser solution.

**Comment by zhangtao2-1** [comment]
Thanks @pwilkin  — I've refactored to the autoparser-first approach: removed the dedicated MiniCPM5 parser, extended autoparser for direct <function name="..."> calls with call_separator, and added a small diff-analyzer workaround only where needed. Local tests pass

**Comment by JINZIPING** [comment]
I retested latest `829d6f6`. MiniCPM5 XML tool parsing still has a partial/streaming edge case.

  For output like:

  ```xml
  <function name="category_menu"><param name="category">Dessert</param></function>
```

  I can still get args shaped like:

  `{"_raw_arguments":"{}\"category\":\"Dessert\"}"}`

  It looks like the partial PEG mapper finalizes the initial { placeholder into {} before the first <param> arrives, then later arg deltas append after {}.

  A narrow local fix was:

  - treat target == "{}" as an empty placeholder when the first arg name is parsed
  - in partial parsing, do not close braces for an empty tool-arg placeholder before any arg has been seen

  I verified this locally with test-chat and test-chat-peg-parser.

**Comment by pwilkin** [comment]
Is the short syntax really a thing? Would it hurt to constrain the model to just the normal form? I'm not a huge fan of adding the `alt` markers to autoparser because of the way they're implemented here - the autoparser is supposed to rely on automatic detection mechanisms and here the detection is impossible.

**Comment by zhangtao2-1** [comment]
Hi @pwilkin @JINZIPING  Follow-up pushed: streaming {} placeholder fix (per JINZIPING), removed alt/compact XML markers (per pwilkin), plus a category_menu regression test. test-chat / test-chat-peg-parser pass. PTAL.

**Comment by zhangtao2-1** [comment]
Hi @pwilkin @CISC @aldehir 
Can we run CI tests next?

**Comment by zhangtao2-1** [comment]
Hi @CISC Pushed 809b392 to fix test-jinja-py CI failures (skip min/max attribute tests in -py mode). Could a maintainer please Approve and run the pending workflows? Thanks!

File: tests/test-jinja.cpp [hunk 19]
```diff
@@ -1534,6 +1534,39 @@ static void test_array_methods(testing & t) {
...
+
+    test_template(t, "array|max",
+        "{{ [tool_calls_count, tool_sep_count]|max }}",
+        {{"tool_calls_count", 2}, {"tool_sep_count", 1}},
+        "2"
+    );
+
+    // attribute= is not implemented in C++ yet; skip in -py mode (Python Jinja2 renders output)
+    if (!g_python_mode) {
```
Comment (CISC) [inline tests/test-jinja.cpp]: Don't do this, restore the tests unconditionally, but put the real expected output instead of `""`.

Comment (zhangtao2-1) [reply tests/test-jinja.cpp]: restored attribute tests with real expected output ({'x': 1} / {'x': 2}). Could you approve the pending workflows when you have a moment? Thanks!

**Comment by zhangtao2-1** [comment]
Hi @CISC 
The gpu-openvino-low-perf failure looks unrelated to this PR — OpenVINO setup failed while downloading/extracting the toolkit (gzip: stdin: not in gzip format). The download URL works when tested manually, so this is likely a transient network issue on the self-hosted runner. Could please re-run the failed job? Thanks!

**Comment by zhangtao2-1** [comment]
Hi @pwilkin @aldehir Refactored to autoparser-first per earlier feedback; streaming fix and jinja tests are in. CI passes. Would appreciate a re-review when you have time. Thanks!

**Comment by aldehir** [comment]
@zhangtao2-1 exactly why do we need to maintain multiple calling notations? Can we not leverage the grammar sampler to ensure the correct notation? 

Please help me understand. As it stands, there is a lot of added complexity here just to support this.

**Comment by zhangtao2-1** [comment]
@aldehir  Good question.

MiniCPM5 only uses one notation: <function name="..."><param name="...">...</param></function>. Alt/compact XML markers were removed per @pwilkin's feedback.

Grammar sampling is already in place — autoparser builds GBNF with a <function trigger. The remaining changes aren't for multiple notations:

diff-analyzer workaround — MiniCPM5 has no outer tool-call wrapper; calls start directly at <function
incomplete-XML / streaming fixes — handle partial output before closing tags arrive (server streaming)
history JSON fallback — malformed tool_calls.arguments in multi-turn agent history, unrelated to generation
Happy to trim unused generic alt/compact infrastructure if you think that's still too much scope for this PR.

**Comment by zhangtao2-1** [comment]
Hi @aldehir — just checking in. Any remaining concerns on your side, or is this ready for another look? Thanks!

**Comment by aldehir** [comment]
> incomplete-XML / streaming fixes — handle partial output before closing tags arrive (server streaming)
history JSON fallback — malformed tool_calls.arguments in multi-turn agent history, unrelated to generation

Please revert these. When implementing model support we need to be risk averse and not unnecessarily change logic that is used across all models to satisfy a single one.

The parser should handle incomplete output if terminated midway, if it doesn't then that might indicate a parser problem.

As for incomplete arguments, ideally the client should not include an incomplete assistant message, and the parser + grammar constrain decoding should ensure that a complete tool call is generated.

We're dragging on merging this due to the liberties taken here and should instead strive to minimize impact in shared logic.

As I understand it, the tool calling notation is similar to Qwen3 Coder without the enclosing tool call tag and with an additional wrapper around string arguments (`CDATA`). We can hardcode these in the diff analysis, as you've done, and possibly introduce a new pair of string argument markers to include in the parser generation. Unless I'm mistaken, that should be the minimum required to support the model.

**Comment by zhangtao2-1** [comment]
Hi @aldehir 
Reverted the shared mapper streaming workarounds and the malformed tool-call history fallback in 7604475cd, per your feedback. MiniCPM5 support is now autoparser/diff-analyzer only. PTAL.

**Comment by zhangtao2-1** [comment]
Hi @aldehir  — I've reverted the shared mapper streaming workarounds and the history JSON fallback as you requested (7604475). MiniCPM5 support is now limited to the autoparser + diff-analyzer only, no changes to shared logic. Could you take another look when you have a moment? Also, would appreciate it if a maintainer could kick off the pending CI workflows. Thanks!

**Comment by aldehir** [comment]
@zhangtao2-1 please review #24889
